File: 0210_Course Schedule II/CourseScheduleII.java
class Solution {
    
    private final int VISITED = 1;
    private final int VISITING = 2;
    
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        
        int[] visited = new int[numCourses];
        List<List<Integer>> graph = new ArrayList<List<Integer>>();
        List<Integer> result = new LinkedList<Integer>();
        
        //initialize a graph
        for(int i=0;i<numCourses;i++){
            graph.add(new ArrayList<Integer>());
        }
        
        //build a graph
        for(int i=0;i<prerequisites.length;i++){
            int course = prerequisites[i][0];
            int preCourse = prerequisites[i][1];
            graph.get(preCourse).add(course);
        }
        
        for(int i=0;i<numCourses;i++){
            if(hasCircle(graph, visited, i, result)) return new int[0];
        }
        
        return result.stream().mapToInt(i->i).toArray();
     
    }
    
    //return true if there is a circle.
    private boolean hasCircle(List<List<Integer>> graph, int[] visited, int curr, List<Integer> result){
        if(visited[curr] == VISITED) return false;
        if(visited[curr] == VISITING) return true;
        visited[curr] = VISITING;
        
        for(int next : graph.get(curr)){
            if(hasCircle(graph, visited, next,result)) return false;
        }
        
        result.add(0,curr);
        visited[curr] = VISITED;
        return false;
    }
}

File: 0210_Course Schedule II/CourseScheduleII.cs
public class Solution {
    private const int VISITING = 1;
    private const int VISITED = 2;
    
    public int[] FindOrder(int numCourses, int[][] prerequisites) {
        var graph = new List<List<int>>(numCourses);
        var visited = new int[numCourses];
        
        for(int i=0;i<numCourses;i++){
            graph.Add(new List<int>());
        }
        
        foreach(var item in prerequisites){
            var course = item[0];
            var preCourse = item[1];
            graph[preCourse].Add(course);
        }
        
        var stack = new Stack<int>(numCourses);
        for(int course = 0; course < numCourses; course++)
            if(DFS(graph, course, visited, stack)) return new int[0];
        
        int[] ans = new int[numCourses];
    
        for(int i=0;i< numCourses;i++){
            ans[i] = stack.Pop();
        }
        
        return ans;
    }
    
    private bool DFS(List<List<int>> graph, int v, int[] visited, Stack<int> stack){
        if(visited[v] == VISITED) return false;
        if(visited[v] == VISITING) return true;
        visited[v] = VISITING;
        foreach(var i in graph[v]){
            if(DFS(graph, i, visited, stack)) return true;
        }
        
        visited[v] = VISITED;
        stack.Push(v);
        return false;
    }
    
}